### ISSUE TITLE
notebook 5, padding in conv

### ISSUE THREAD

**Issue by antgr** [bentrevett/pytorch-seq2seq#30]
In encoder, we include padding inside the definition of each nn.Conv1d, 
`self.convs = nn.ModuleList([nn.Conv1d(in_channels = hid_dim, 
                                              out_channels = 2 * hid_dim, 
                                              kernel_size = kernel_size, 
                                              **padding** = (kernel_size - 1) // 2)
                                    for _ in range(n_layers)])`
but in decoder, we don't 
`self.convs = nn.ModuleList([nn.Conv1d(hid_dim, 2*hid_dim, kernel_size)
                                    for _ in range(n_layers)])`
and we work around this, with the following code:
`#need to pad so decoder can't "cheat"
            padding = torch.zeros(conv_input.shape[0], conv_input.shape[1], self.kernel_size-1).fill_(self.pad_idx).to(device)
            padded_conv_input = torch.cat((padding, conv_input), dim=2) `
Why do we follow different treatment, between those two cases?
Thanks again!

**Comment by bentrevett** [comment]
When encoding, we don't care if the encoder sees all of the tokens within a sentence. If anything, we want that to happen so it's encoding of a word will be influenced by the other words in the sentence. We only pad in the decoder so that the length of the sentence stays the same through each layer. 

When decoding, we are passing in the whole of the target sentence all at once as we want to process it all in parallel. If we let the decoder see the whole sentence, then it won't actually learn how to translate, it will just learn to look at the word in the next position and just output that. So, we have to pad the sentence - at the beginning of the word - so that each filter cannot see "future" tokens which we are trying to predict, i.e. the first filter can only see the first token, the second can only see the second token and previous tokens, etc. 

This is why we add padding of `filter_size - 1` to the start of each target sentence in the decoder.

This image might also be useful:

![](https://i.imgur.com/Zd2Tdmd.jpg)

I am planning to hopefully finish notebook 5 soon(ish) and will add all of this to the tutorial.

**Comment by antgr** [comment]
So, we have two kind of paddings one for the encoder and a different one for the decoder. 
1) for the encoder is the usual padding for conv1d, to keep the size the same, and 
2) for the decoder is for the reason you analyzed: to NOT let the decoder see the whole sentence at each step (where "step" here does not literally exist, because we do the processing in parallel).
I will look again into the code, searching in it the details of your answer, and to fully clarify everything.
Thanks again for your valuable help !! :)

**Comment by bentrevett** [comment]
Yep, you're correct on both. Glad I could help.
